### PR TITLE
Feat: Activation Checkpointing

### PR DESCRIPTION
**Pull request by Chamberlain0w0**
## 背景

本 PR 为 InfiniTrain 增加重计算能力，通过在反向传播时重新执行部分前向计算，减少训练过程中需要保留的中间激活，以计算开销换取显存占用下降。

整体配置与 Megatron-Core 的 Transformer activation recompute 语义保持一致。

### 通用 Checkpoint 接口

新增 `utils::checkpoint::Checkpoint()`，对标 [`torch.utils.checkpoint()`](https://docs.pytorch.ac.cn/docs/stable/checkpoint.html)。同时支持：

- Reentrant checkpoint
- Non-reentrant checkpoint，默认使用该模式

Non-reentrant 模式通过 Saved Tensor Hooks 将前向过程中保存的中间 Tensor 替换为占位对象。反向传播真正读取这些 Tensor 时，再按需重新执行 checkpoint 区域的前向计算。

该实现包含以下机制：

- Early stop：需要的中间结果全部生成后，提前终止重计算
- Determinism check：校验重计算结果的 shape、dtype 和 device 是否与原始前向一致
- Autocast context 恢复：重计算沿用原始前向的 autocast 配置，保证 BF16 等混合精度场景的一致性
- 输入 Tensor detach 后重建 `requires_grad` 状态，避免错误连接原始计算图
- 重计算结束后及时释放 Tensor 和引用，降低峰值显存并避免引用环

由于当前 autograd engine 在构图时会直接修改依赖计数，non-reentrant 重计算在 `no_grad` 下执行，并新增 `PropagateRequiresGradGuard` 传递 `requires_grad` 和 `needs_input_grad` 元数据，避免临时重计算图污染原始反向图。

### Transformer 重计算策略

`TransformerConfig` 新增以下配置：

- `recompute_granularity`
  - `none`：关闭重计算
  - `full`：以完整 Transformer Layer 为重计算单位
  - `selective`：预留配置，当前尚未实现
- `recompute_method`
  - `none`：每个 Transformer Layer 独立 checkpoint
  - `uniform`：每连续 `recompute_num_layers` 层组成一个 checkpoint 区域
  - `block`：仅 checkpoint 当前本地 chunk/stage 的前 `recompute_num_layers` 层
- `recompute_num_layers`：控制 uniform 分组大小或 block 覆盖层数

未启用重计算或处于 `no_grad` 场景时，Transformer 仍直接执行原有前向路径。

## 主要修改

- 新增通用 activation checkpoint API，以及 reentrant/non-reentrant 实现
- 扩展 autograd grad mode，使 non-reentrant 重计算可以传递梯度相关元数据而不构建临时计算图
- 在 Transformer Layer 执行路径中接入 full activation recompute
- 为 GPT-2 和 LLaMA 3 示例新增命令行参数：
  - `--activation_recompute`
  - `--recompute_granularity`
  - `--recompute_method`
  - `--recompute_num_layers`
- LLMC checkpoint loader 支持将运行时 recompute 配置应用到加载后的模型
- 统一 Module 调用入口，确保前向 hooks 等 Module 调用协议能够正常执行
- 将 MNIST 示例中的直接 `Forward()` 调用调整为 `operator()`

## 当前限制

- `selective` granularity 尚未实现，目前仅支持完整 Transformer Layer 重计算
- RNG state 保存与恢复尚未实现，因此 checkpoint 区域暂不支持依赖随机状态一致性的 Dropout 或其他随机算子
- Reentrant checkpoint 当前仅支持单输出 forward

### CODE REVIEW

File: example/gpt2/checkpoint_loader.h [hunk 1]
```diff
@@ -3,10 +3,14 @@
...
+#include "infini_train/include/nn/modules/transformer/transformer_config.h"
+
 namespace infini_train::nn {
 class TransformerModel;
 } // namespace infini_train::nn
 
 namespace gpt2 {
 std::shared_ptr<infini_train::nn::TransformerModel> LoadFromLLMC(const std::string &filepath);
+std::shared_ptr<infini_train::nn::TransformerModel>
+LoadFromLLMC(const std::string &filepath, const infini_train::nn::TransformerConfig &runtime_config);
```
Comment (chen2021673) [inline example/gpt2/checkpoint_loader.h]: 这个接口需要吗，namespace gpt2下直接在LoadFromLLMC里写死gpt2::GPT2Config()行不。另外现在的逻辑是如果用户传入了 llmc_path，应该拿 .bin 里存储的 config 信息重新修正 model_config，会覆盖这里的runtime_config吗

File: example/gpt2/checkpoint_loader.cc [hunk 2]
```diff
@@ -52,11 +52,22 @@ std::tuple<int32_t, infini_train::DataType> DetermineAndCheckVersion(const std::
         return {}; // Unreachable, but keeps compiler happy
     }
 }
+
+void ApplyRuntimeRecomputeConfig(nn::TransformerConfig *config, const nn::TransformerConfig &runtime_config) {
```
Comment (chen2021673) [inline example/gpt2/checkpoint_loader.cc]: GPT2/LLaMA3 里都有这个重复的 ApplyRuntimeRecomputeConfig，能不能抽成公共的。另外参数要不要用 CheckpointOptions 包一下

File: infini_train/src/utils/checkpoint.cc [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    saved_inputs_.clear();
+    saved_inputs_requires_grad_.clear();
+    return grad_inputs;
+}
+
+std::vector<std::shared_ptr<Tensor>> Checkpoint(const CheckpointFunction::ForwardFn &forward_fn,
+                                                const std::vector<std::shared_ptr<Tensor>> &inputs, bool use_reentrant,
+                                                bool preserve_rng_state, bool determinism_check, bool early_stop) {
+    if (preserve_rng_state) {
+        // TODO(zbl): Preserve and restore RNG state for CPU/CUDA.
```
Comment (chen2021673) [inline infini_train/src/utils/checkpoint.cc:278]: 这里留一个 LOG(FATAL) 吧

File: infini_train/include/utils/checkpoint.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+#include "infini_train/include/autocast.h"
+#include "infini_train/include/autograd/function.h"
+
+namespace infini_train {
+class Tensor;
+} // namespace infini_train
+
+namespace infini_train::utils::checkpoint {
+
+class CheckpointFunction : public autograd::Function {
```
Comment (chen2021673) [inline infini_train/include/utils/checkpoint.h]: 这个类放在utils里了，但总感觉不应该作为外部 API 暴露出去，要不加到匿名 namespace 里？

File: infini_train/src/nn/modules/transformer/transformer.cc [hunk 5]
```diff
@@ -1,12 +1,15 @@
 #include "infini_train/include/nn/modules/transformer/transformer.h"
```
Comment (chen2021673) [inline infini_train/src/nn/modules/transformer/transformer.cc:1]: 之后扩展 recompute 其他功能（如selective）的话就不适合全放在transformer.cc里了，之后可以拆分activation_recompute.cc。本PR可以不做修改

File: infini_train/src/utils/checkpoint.cc [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+            frame->inputs.push_back(input->Detach());
+            frame->inputs_requires_grad.push_back(input->requires_grad());
+        } else {
+            frame->inputs.push_back(nullptr);
+            frame->inputs_requires_grad.push_back(false);
+        }
+    }
+    frame->autocast_state = GetAutocastState();
+
+    autograd::Function::SavedTensorHooks hooks;
```
Comment (chen2021673) [inline infini_train/src/utils/checkpoint.cc]: 在有两套 SavedTensorHooks 的情况下，这个hooks能不能改名叫placeholder_hooks，方便理解一点。

File: infini_train/include/autocast.h [hunk 7]
```diff
@@ -164,6 +164,24 @@ struct AutocastContext {
...
+    Device::DeviceType device_type = Device::DeviceType::kCPU;
+    DataType autocast_dtype = DataType::kBFLOAT16;
+};
+
+inline AutocastState GetAutocastState() {
+    return AutocastState{tls_autocast_context.enabled, tls_autocast_context.device_type,
+                         tls_autocast_context.autocast_dtype};
+}
+
+inline void SetAutocastState(const AutocastState &state) {
```
Comment (chen2021673) [inline infini_train/include/autocast.h]: 这个SetAutocastState能不能也写成guard

File: infini_train/src/nn/modules/module.cc [hunk 8]
```diff
@@ -152,7 +151,8 @@ std::vector<std::shared_ptr<Tensor>> Module::Forward(const std::vector<std::shar
     return {};
 }
 
-std::vector<std::shared_ptr<Tensor>> Module::operator()(const std::vector<std::shared_ptr<Tensor>> &input_tensors) {
+std::vector<std::shared_ptr<Tensor>>
+Module::ForwardWithHooks(const std::vector<std::shared_ptr<Tensor>> &input_tensors) {
```
Comment (chen2021673) [inline infini_train/src/nn/modules/module.cc]: 这个能不能叫Module::Call，因为我理解这里才是通用入口，对应 PyTorch 里的 Module.__call__ 语义

File: infini_train/src/utils/checkpoint.cc (hunk 3 above)
Comment (Chamberlain0w0) [reply infini_train/src/utils/checkpoint.cc:278]: ok，改了

File: example/gpt2/checkpoint_loader.cc (hunk 2 above)
Comment (Chamberlain0w0) [reply example/gpt2/checkpoint_loader.cc]: 抽出来了，但是先叫做 ActivationRecomputeOptions，感觉 CheckpointOptions 容易和模型 ckpt 混淆。

File: infini_train/include/utils/checkpoint.h (hunk 4 above)
Comment (Chamberlain0w0) [reply infini_train/include/utils/checkpoint.h]: 有道理，改了

File: infini_train/src/nn/modules/transformer/transformer.cc (hunk 5 above)
Comment (Chamberlain0w0) [reply infini_train/src/nn/modules/transformer/transformer.cc:1]: 确实，这块确实没太想清楚，主要是 megatron 的实现里面也把很多重计算逻辑融在了 transformer 模型层，之后可以再讨论下

File: infini_train/src/utils/checkpoint.cc (hunk 6 above)
Comment (Chamberlain0w0) [reply infini_train/src/utils/checkpoint.cc]: ok，改了

File: infini_train/include/autocast.h (hunk 7 above)
Comment (Chamberlain0w0) [reply infini_train/include/autocast.h]: ok 改了

File: infini_train/src/nn/modules/module.cc (hunk 8 above)
Comment (Chamberlain0w0) [reply infini_train/src/nn/modules/module.cc]: 嗯嗯，改了

File: example/gpt2/checkpoint_loader.h (hunk 1 above)
Comment (Chamberlain0w0) [reply example/gpt2/checkpoint_loader.h]: 其实我想的是传入 runtime config 而已，之前和现在的 namespace gpt2 下确实是直接在 LoadFromLLMC 里写死 gpt2::GPT2Config() 的，这里传入的只会针对 runtime config 的部分做个更新。考虑到后面的修改提出了一个ActivationRecomputeOptions，我这里就直接用 ActivationRecomputeOptions 类型作为参数传入了，避免混淆。

File: infini_train/include/autograd/grad_mode.h [hunk 9]
```diff
@@ -34,4 +37,19 @@ class EnableGradGuard {
     bool prev_;
 };
 
+// RAII: Propagate requires_grad metadata while graph construction is disabled.
+// Used by non-reentrant checkpoint recomputation so downstream SetupContext
+// calls see the same needs_input_grad_ pattern as the original forward,
+// without wiring the recompute graph into the engine.
+class PropagateRequiresGradGuard {
```
Comment (Chamberlain0w0) [inline infini_train/include/autograd/grad_mode.h:44]: 本质上是需要一个 2*2=4 种情况的精细控制（with_grad/no_grad x 带引用计数建图/仅传递 requires_grad）。

也可以不单独实现一个 guard，而是给原先的 GradGuard 重载一个带参数 (比如 bool record_context = false) 的构造方式，让 GradGuard 本身能够覆盖这四种情况。

File: infini_train/src/utils/checkpoint.cc [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+        };
+        recompute_hooks.unpack = [](const std::shared_ptr<void> &state) -> std::shared_ptr<Tensor> {
+            return std::static_pointer_cast<Tensor>(state);
+        };
+        autograd::FunctionCtx::SavedTensorHooksGuard guard(std::move(recompute_hooks));
+
+        try {
+            forward_fn(detached_inputs);
+        } catch (const StopRecomputeError &) {
+            // Early-stop: expected when all needed tensors are recomputed.
```
Comment (Chamberlain0w0) [inline infini_train/src/utils/checkpoint.cc:181]: 只是用异常的语法和控制流机制来实现 early stop 的语义，本质上并不是真的抛一个会让程序 abort 的异常。torch 也采取的是一样的实现方式。

Image

这块暂时没有想到其他更优雅的处理方式。
